logging: Expand env vars in logging config

logston opened the pull request:
This commit adds a command line flag `--log-config-vars` that is used
  to specify those environment variables that will be expanded in the
  logging config file contents. This flag enables operators to
  indirectly define an allow list of sorts of trusted environment
  variables.

  A desired use case of this flag is one where `HOST_IP` is pulled from
  the running container environment and used to direct logs to a node
  local instance of FluentBit. In this snipped below, the log.yaml
  configuration file can be defined with a generic `fluent-server` sink
  that specifies `${HOST_IP}:5170` as its address. CRDB will then pull
  the IP from the environment and replace the variable with the IP in
  the logging configuration.

  ```
  containers:
    - name: cockroach
      image: cockroachdb/cockroach
      args:
      - --log-config-file log.yaml
      - --log-config-vars HOST_IP
      env:
        - name: HOST_IP
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
  ```

  Closes #81146.

logston committed:
logging: Expand env vars in logging config

  This commit adds a command line flag `--log-config-vars` that is used
  to specify those environment variables that will be expanded in the
  logging config file contents. This flag enables operators to
  indirectly define an allow list of sorts of trusted environment
  variables.

  A desired use case of this flag is one where `HOST_IP` is pulled from
  the running container environment and used to direct logs to a node
  local instance of FluentBit. In this snippet below, the log.yaml
  configuration file can be defined with a generic `fluent-server` sink
  that specifies `${HOST_IP}:5170` as its address. CRDB will then pull
  the IP from the environment and replace the variable with the IP in
  the logging configuration.

  ```
  containers:
    - name: cockroach
      image: cockroachdb/cockroach
      args:
      - --log-config-file log.yaml
      - --log-config-vars HOST_IP
      env:
        - name: HOST_IP
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
  ```

  Closes #81146.

Release note (cli change): The CLI now contains a flag
(--log-config-vars) that allows for environment variables to be
specified for expansion within the logging configuration file. This
change allows for a single logging configuration file to service an
array of sinks without further manipulation of the configuration file.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82147)
<!-- Reviewable:end -->

logston commented:
I've just pushed up this draft solution for issue #81146. This PR doesn't have tests, docs, etc as I'd like to get consensus on direction before I implement more. Please chime in if you have suggestions on implementation.

knz reviewed (commented):
I like this.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-:-N3PpNjWEwS00u9ht2tC:b-rlob0d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @logston)

___
*[`pkg/cli/flags.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Pohj4D0Um6Od36Rex:-N3Pohj5AXdemwTRhvJl:bomwxkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/flags.go#L368)):*
> ```Go
> 		varFlag(pf, &stringValue{settableString: &cliCtx.logConfigInput}, cliflags.Log)
> 		varFlag(pf, &fileContentsValue{settableString: &cliCtx.logConfigInput, fileName: "<unset>"}, cliflags.LogConfigFile)
> 		varFlag(pf, &sliceValue{settableSlice: &cliCtx.logConfigVars}, cliflags.LogConfigVars)
> ```

Please use `stringSliceFlag`, you don't need to define `settableSlice`.
___
*[`pkg/cli/log_flags.go` line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Poo7E58oOTjpW_jvf:-N3Poo7E58oOTjpW_jvg:b-2lhiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/log_flags.go#L515)):*
> ```Go
> 	for _, v := range vars {
> 		if strings.HasPrefix(strings.ToUpper(v), "COCKROACH_") {
> 			return errors.Newf("use of %s is not allowed as a logging configuration variables", v)
> ```

nit: variables -> variable
___
*[`pkg/cli/log_flags.go` line 527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Pp1sb1Vnf2uhtphR6:-N3Pp1sb1Vnf2uhtphR7:b-fin1ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/log_flags.go#L527)):*
> ```Go
> 		val := os.Getenv(k)
> 		// Replace $ENV_VAR and ${ENV_VAR} syntaxes.
> 		s = strings.ReplaceAll(s, fmt.Sprintf("$%s", k), val)
> ```

This is unsafe and also somewhat inefficient.
May I instead recommend `os.Extend()`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @logston)

___
*[`pkg/cli/log_flags.go` line 527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Pp1sb1Vnf2uhtphR6:-N3PpSX-1DzMIKoBFZNK:btkbl6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/log_flags.go#L527)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This is unsafe and also somewhat inefficient.
May I instead recommend `os.Extend()`.
</blockquote></details>

I mean `os.Expand()`


<!-- Sent from Reviewable.io -->

logston reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @knz)

___
*[`pkg/cli/flags.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Pohj4D0Um6Od36Rex:-N3Uff_125dKgE5kdVNp:b-59df7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/flags.go#L368)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please use `stringSliceFlag`, you don't need to define `settableSlice`.
</blockquote></details>

Thanks! Will do. 
___
*[`pkg/cli/log_flags.go` line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Poo7E58oOTjpW_jvf:-N3UflG-CFaxNfQ6bZ3u:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/log_flags.go#L515)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: variables -> variable
</blockquote></details>

Thanks.
___
*[`pkg/cli/log_flags.go` line 527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3Pp1sb1Vnf2uhtphR6:-N3UfpUtEDAlI-_yUeyT:b-cza0et) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5be6e29256564f0c01193df8c167eb5ce940fe3/pkg/cli/log_flags.go#L527)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I mean `os.Expand()`
</blockquote></details>

Ah, great. I was hoping some other method existed. 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 3 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp, @knz, and @logston)

___
*[`pkg/cli/context.go` line 227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3V6OjC6v_8o3t9qYot:-N3V6OjC6v_8o3t9qYou:bvfejhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bfabed8f8ad03aa5267952cc3ce997cc7a0435f/pkg/cli/context.go#L227)):*
> ```Go
> 	cliCtx.allowUnencryptedClientPassword = false
> 	cliCtx.logConfigInput = settableString{s: ""}
> 	cliCtx.logConfigVars = []string{}
> ```

more idiomatic go: `= nil`
___
*[`pkg/cli/log_flags.go` line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3V6fLU9lM8AV21wtFX:-N3V6fLU9lM8AV21wtFY:bs56pmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bfabed8f8ad03aa5267952cc3ce997cc7a0435f/pkg/cli/log_flags.go#L499)):*
> ```Go
> 			return v
> 		}
> 		return ""
> ```

Is a silent no-op what we want here wrt UX?

Isn't it desirable to inform the operator if they have e.g. a typo in their variable name?


<!-- Sent from Reviewable.io -->

logston reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @knz)

___
*[`pkg/cli/context.go` line 227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3V6OjC6v_8o3t9qYot:-N3VBDaYAqWrSf845t7o:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bfabed8f8ad03aa5267952cc3ce997cc7a0435f/pkg/cli/context.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

more idiomatic go: `= nil`
</blockquote></details>

:+1: 
___
*[`pkg/cli/log_flags.go` line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3V6fLU9lM8AV21wtFX:-N3V9FCm0PplTUXNoxc9:byzzwy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bfabed8f8ad03aa5267952cc3ce997cc7a0435f/pkg/cli/log_flags.go#L499)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Is a silent no-op what we want here wrt UX?

Isn't it desirable to inform the operator if they have e.g. a typo in their variable name?
</blockquote></details>

Good question. I was wavering back and forth on this and decided to go with a solution that mimics what happens if the var is not defined in the environment. But your right that it's not a great UX. What do you think about a panic here instead? We could return an error in `expandEnvironmentVariables` but the solution I see there is not as clean as I'd like. 


<!-- Sent from Reviewable.io -->

knz reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp, @knz, and @logston)

___
*[`pkg/cli/log_flags.go` line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3V6fLU9lM8AV21wtFX:-N3VBa5j2i94lAqrsWv3:b2kqoqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bfabed8f8ad03aa5267952cc3ce997cc7a0435f/pkg/cli/log_flags.go#L499)):*
<details><summary><i>Previously, logston (Paul Logston) wrote…</i></summary><blockquote>

Good question. I was wavering back and forth on this and decided to go with a solution that mimics what happens if the var is not defined in the environment. But your right that it's not a great UX. What do you think about a panic here instead? We could return an error in `expandEnvironmentVariables` but the solution I see there is not as clean as I'd like. 
</blockquote></details>

```go
var err error
result := os.Expand(s, func(k string) string {
   if !ok {
      err = errors.CombineErrors(err, errors.Newf("variable %q is not defined"))
      return ""
   }
   ...
})
return result, err
```



_Code quote:_
```Go
	return os.Expand(s, func(k string) string {
		v, ok := m[k]
		if ok {
			return v
		}
		return ""
	})
```


<!-- Sent from Reviewable.io -->

logston reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @knz)

___
*[`pkg/cli/log_flags.go` line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3V6fLU9lM8AV21wtFX:-N3VE3y5-EP0EItOA_TC:brv7b4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bfabed8f8ad03aa5267952cc3ce997cc7a0435f/pkg/cli/log_flags.go#L499)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

```go
var err error
result := os.Expand(s, func(k string) string {
   if !ok {
      err = errors.CombineErrors(err, errors.Newf("variable %q is not defined"))
      return ""
   }
   ...
})
return result, err
```
</blockquote></details>

:+1: Forgot about `CombineErrors`. Thanks. 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @logston)

___
*[`pkg/cli/log_flags.go` line 495 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3YOQYc-7R3v8o6J_5H:-N3YOQYc-7R3v8o6J_5I:b-prp20p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea4a3c30115a70ceb137fbfa2e0ac0b8317a8343/pkg/cli/log_flags.go#L495)):*
> ```Go
> 	// in vars to create an allow list.
> 	for _, k := range vars {
> 		m[k] = os.Getenv(k)
> ```

it would be good to add a new accessor in pkg/util/envutil for this, one that adds the env var used to the registry, so that it gets included in env var dumps.

Placing the accessor in envutil will also make the linter happy.
___
*[`pkg/cli/log_flags_test.go` line 235 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3YOj5H63LFh_NeK3vw:-N3YOj5H63LFh_NeK3vx:bufble5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea4a3c30115a70ceb137fbfa2e0ac0b8317a8343/pkg/cli/log_flags_test.go#L235)):*
> ```Go
> 
> func TestExpandEnvironmentVariables(t *testing.T) {
> 	if err := os.Setenv("HOST_IP", "1.2.3.4"); err != nil {
> ```

tip: `require.NoError`, here and below to replace `if err != nil { t.Fatal`


<!-- Sent from Reviewable.io -->

knz commented:
Don't forget somewhere on the way to involve the obs infra team, which ultimately will be responsible to maintain this forward. You need to get buy-in from them first and foremost. You can ask @dhartunian  to guide you to the right place.

logston commented:
Ah, thanks @knz. I'm working remotely without Slack and about to go on vacation, thus this will be my last day working on this for quite a while. @jmcarp or @joshimhoff, I'll need you to pass of context to @dhartunian, if you would please. 

@dhartunian, the TLDR is that we've had a number of stability issues due to the fact that we need to know the node IP on which the CRDB instance is running for the logging configuration. But our current method of placing that node IP in the configuration is not as robust as we'd like. This led us to the idea that some form of env var expansion in config file itself would be an improvement.

joshimhoff commented:
@dhartunian, if you want more info from me, just give me a poke somehow!

logston reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @knz)

___
*[`pkg/cli/log_flags.go` line 495 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3YOQYc-7R3v8o6J_5H:-N3_3S-3BRY3HqCyUvjT:b-fyzhgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea4a3c30115a70ceb137fbfa2e0ac0b8317a8343/pkg/cli/log_flags.go#L495)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

it would be good to add a new accessor in pkg/util/envutil for this, one that adds the env var used to the registry, so that it gets included in env var dumps.

Placing the accessor in envutil will also make the linter happy.
</blockquote></details>

Ah, thank you. 
___
*[`pkg/cli/log_flags_test.go` line 235 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3YOj5H63LFh_NeK3vw:-N3_6U8hCs6rOgy5gjA8:bqhi13u) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea4a3c30115a70ceb137fbfa2e0ac0b8317a8343/pkg/cli/log_flags_test.go#L235)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

tip: `require.NoError`, here and below to replace `if err != nil { t.Fatal`
</blockquote></details>

This seems to break with the convention in other parts of the file. Are there certain circumstances where we wan to fail fast an others were we want a summary of failures? If so, do we have those listed somewhere? 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I'll also chip in and help bring the PR to resolution while Paul is away if needs be.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-:-N3_E8V23ZmbFujvVH4P:b-vza8c3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jmcarp and @logston)

___
*[`pkg/cli/log_flags_test.go` line 235 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N3YOj5H63LFh_NeK3vw:-N3_DghG5DSzkuUcfkzI:bzadba) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea4a3c30115a70ceb137fbfa2e0ac0b8317a8343/pkg/cli/log_flags_test.go#L235)):*
> This seems to break with the convention in other parts of the file

Most of the CLI testing code pre-dates our collective discovery of the stretchr/require and stretchr/assert packages. We've started to introduce their use in new test code. Hence my suggestion.

> . Are there certain circumstances where we wan to fail fast an others were we want a summary of failures?

`require.NoError` does t.Fatal under the hood (fail fast);
`assert.NoError` does t.Error (summary at end).

The choices you've already made of t.Fatal/t.Error look good.

>  If so, do we have those listed somewhere?
I don't think so.


<!-- Sent from Reviewable.io -->

logston committed:
util:envutil: Add external env var accessor

  This commit adds a method `ExternalEnvString` that can be used for
  fetching explicitly non-CRDB environment variables. These variables
  are stored in the same cache as used CRDB environment variables but
  are guaranteed to not be prefixed with `COCKROACH_`

Release note (<category, see below>): <what> <show> <why>

dhartunian reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just need some test tweaks. i think on rebase some of the outputs changed in `TestSetupLogging/server_overrides`, so worth updating them.

> Reviewed 4 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-:-N7XnkCT1veye1OcfSce:bxwuetx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @jmcarp, @jreut, @knz, and @logston)

___
*[`pkg/cli/log_flags_test.go` line 222 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N7XoD0F0xDb5yRPiJp3:-N7XoD0F0xDb5yRPiJp4:b-6fmlsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bb3359a30cb2b9f57e9df0a104a7b7b0cdfca8b/pkg/cli/log_flags_test.go#L222)):*
> ```Go
> }
> 
> func TestValidateLogConfigVars(t *testing.T) {
> ```

add `defer leaktest.AfterTest(t)()`
___
*[`pkg/cli/log_flags_test.go` line 244 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N7XoFV5-vavm0Whm2MI:-N7XoFV5-vavm0Whm2MJ:b-6fmlsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bb3359a30cb2b9f57e9df0a104a7b7b0cdfca8b/pkg/cli/log_flags_test.go#L244)):*
> ```Go
> }
> 
> func TestExpandEnvironmentVariables(t *testing.T) {
> ```

add `defer leaktest.AfterTest(t)()`
___
*[`pkg/cli/log_flags_test.go` line 286 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/82147#-N7Xn44P8Ecfa6q0tbzW:-N7Xn44P8Ecfa6q0tbzX:b13g87m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bb3359a30cb2b9f57e9df0a104a7b7b0cdfca8b/pkg/cli/log_flags_test.go#L286)):*
> ```Go
> 		},
> 	} {
> 		out, err := expandEnvironmentVariables(tc.in, tc.vars)
> ```

This is failing because of code in `https://github.com/cockroachdb/cockroach/blob/84f351f326af343b9d1b20f4083587ed52299f5f/pkg/util/envutil/env.go#L72-L77 that checks for env var retrieval from different call sites.

I think if you just don't repeat env vars in the test, you should be fine.


<!-- Sent from Reviewable.io -->

logston commented:
bors r+
